<launch>

	<arg name="setpoint" default="true" />

	<node name="pose_pid" pkg="motor_controllers" type="Pose_PID.py" output="screen"/>
	<node name="pid_pololu" pkg="motor_controllers" type="PID_Pololu_matrix.py" output="screen"/>

	<node pkg="topic_tools" type="mux" name="depth_mux" ns="cusub_common/motor_controllers" args="mux/depth/control_effort pid/depth/control_effort cv/depth/control_effort">
		<remap from="mux" to="depth_mux"/>
		<param name="initial_topic " value="cusub_common/motor_controllers/pid/depth/control_effort" />
	</node>
	<node name="depth_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers" >
		<param name="node_name" value="depth_pid" />
		<param name="topic_from_controller" value="pid/depth/control_effort" />
		<param name="topic_from_plant" value="pid/depth/state" />
		<param name="setpoint_topic" value="pid/depth/setpoint" />
		<param name="Kp" value="-100.0" />
		<param name="Ki" value="0.0" />
		<param name="Kd" value="-50.0" />
		<param name="rate" value="25" />
		<param name="upper_limit" value="100" />
		<param name="lower_limit" value="-100" />
		<param name="windup_limit" value=".001" />
		<param name="diagnostic_period" value="0.25" />
		<param name="max_loop_frequency" value="25.0" />
		<param name="min_loop_frequency" value="25.0" />
	</node>
	<node name="depth_cv" pkg="motor_controllers" type="bangbang.py" ns="cusub_common/motor_controllers">
		<param name="setpoint_topic" value="cv/depth/setpoint" />
		<param name="state_topic" value="cv/depth/state" />
		<param name="control_effort_topic" value="cv/depth/control_effort" />
		<param name="deadzone_size" value="0" />
		<param name="left_bang_effort" value="65" />
		<param name="right_bang_effort" value="75" />
	</node>

	<group if="$(arg setpoint)">

		<node name="pitch_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers" >
			<param name="node_name" value="pitch_pid" />
			<param name="topic_from_controller" value="pid/pitch/control_effort" />
			<param name="topic_from_plant" value="pid/pitch/state" />
			<param name="setpoint_topic" value="pid/pitch/setpoint" />
			<param name="Kp" value="20.0" />
			<param name="Ki" value="0.0" />
			<param name="Kd" value="0.0" />
			<param name="rate" value="25" />
			<param name="upper_limit" value="100" />
			<param name="lower_limit" value="-100" />
			<param name="windup_limit" value=".001" />
			<param name="diagnostic_period" value="0.25" />
			<param name="max_loop_frequency" value="25.0" />
			<param name="min_loop_frequency" value="25.0" />
		</node>
		<node name="roll_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers" >
			<param name="node_name" value="roll_pid" />
			<param name="topic_from_controller" value="pid/roll/control_effort" />
			<param name="topic_from_plant" value="pid/roll/state" />
			<param name="setpoint_topic" value="pid/roll/setpoint" />
			<param name="Kp" value="20.0" />
			<param name="Ki" value="0.0" />
			<param name="Kd" value="0.0" />
			<param name="rate" value="25" />
			<param name="upper_limit" value="100" />
			<param name="lower_limit" value="-100" />
			<param name="windup_limit" value=".001" />
			<param name="diagnostic_period" value="0.25" />
			<param name="max_loop_frequency" value="25.0" />
			<param name="min_loop_frequency" value="25.0" />
		</node>
		<node pkg="topic_tools" type="mux" name="yaw_mux" ns="cusub_common/motor_controllers" args="mux/yaw/control_effort pid/yaw/control_effort cv/yaw/control_effort">
			<remap from="mux" to="yaw_mux"/>
			<param name="initial_topic " value="cusub_common/motor_controllers/pid/yaw/control_effort" />
		</node>
		<node name="yaw_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers" >
			<param name="node_name" value="yaw_pid" />
			<param name="topic_from_controller" value="pid/yaw/control_effort" />
			<param name="topic_from_plant" value="pid/yaw/state" />
			<param name="setpoint_topic" value="pid/yaw/setpoint" />
			<param name="Kp" value="500.0" />
			<param name="Ki" value="250.0" />
			<param name="Kd" value="0.0" />
			<param name="rate" value="25" />
			<param name="upper_limit" value="100" />
			<param name="lower_limit" value="-100" />
			<param name="windup_limit" value=".001" />
			<param name="diagnostic_period" value="0.25" />
			<param name="max_loop_frequency" value="25.0" />
			<param name="min_loop_frequency" value="25.0" />
			<param name="angle_error" value="true" />
		</node>
		<node name="yaw_cv" pkg="motor_controllers" type="bangbang.py" ns="cusub_common/motor_controllers">
			<param name="setpoint_topic" value="cv/yaw/setpoint" />
			<param name="state_topic" value="cv/yaw/state" />
			<param name="control_effort_topic" value="cv/yaw/control_effort" />
			<param name="deadzone_size" value="30" />
			<param name="left_bang_effort" value="40" />
			<param name="right_bang_effort" value="-40" />
		</node>
		<node pkg="topic_tools" type="mux" name="drive_mux" ns="cusub_common/motor_controllers" args="mux/drive/control_effort pid/drive/control_effort cv/drive/control_effort">
			<remap from="mux" to="drive_mux"/>
			<param name="initial_topic " value="cusub_common/motor_controllers/pid/drive/control_effort" />
		</node>
		<node name="drive_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers">
			<param name="node_name" value="drive_pid" />
			<param name="topic_from_controller" value="pid/drive/control_effort" />
			<param name="topic_from_plant" value="pid/drive/state" />
			<param name="setpoint_topic" value="pid/drive/setpoint" />
			<param name="Kp" value="30.0" />
			<param name="Ki" value="0.0" />
			<param name="Kd" value="0.0" />
			<param name="rate" value="25" />
			<param name="upper_limit" value="100" />
			<param name="lower_limit" value="-100" />
			<param name="windup_limit" value=".001" />
			<param name="diagnostic_period" value="0.25" />
			<param name="max_loop_frequency" value="25.0" />
			<param name="min_loop_frequency" value="25.0" />
		</node>
		<node name="drive_cv" pkg="motor_controllers" type="bangbang.py" ns="cusub_common/motor_controllers">
			<param name="setpoint_topic" value="cv/drive/setpoint" />
			<param name="state_topic" value="cv/drive/state" />
			<param name="control_effort_topic" value="cv/drive/control_effort" />
			<param name="deadzone_size" value="30" />
			<param name="left_bang_effort" value="60" />
			<param name="right_bang_effort" value="-60" />
		</node>
		<node pkg="topic_tools" type="mux" name="strafe_mux" ns="cusub_common/motor_controllers" args="mux/strafe/control_effort pid/strafe/control_effort cv/strafe/control_effort">
			<remap from="mux" to="strafe_mux"/>
			<param name="initial_topic " value="cusub_common/motor_controllers/pid/strafe/control_effort" />
		</node>
		<node name="strafe_pid" pkg="pid" type="controller" ns="cusub_common/motor_controllers">
			<param name="node_name" value="strafe_pid" />
			<param name="topic_from_controller" value="pid/strafe/control_effort" />
			<param name="topic_from_plant" value="pid/strafe/state" />
			<param name="setpoint_topic" value="pid/strafe/setpoint" />
			<param name="Kp" value="50.0" />
			<param name="Ki" value="0.0" />
			<param name="Kd" value="0.0" />
			<param name="rate" value="25" />
			<param name="upper_limit" value="100" />
			<param name="lower_limit" value="-100" />
			<param name="windup_limit" value=".001" />
			<param name="diagnostic_period" value="0.25" />
			<param name="max_loop_frequency" value="25.0" />
			<param name="min_loop_frequency" value="25.0" />
		</node>
		<node name="strafe_cv" pkg="motor_controllers" type="bangbang.py" ns="cusub_common/motor_controllers">
			<param name="setpoint_topic" value="cv/strafe/setpoint" />
			<param name="state_topic" value="cv/strafe/state" />
			<param name="control_effort_topic" value="cv/strafe/control_effort" />
			<param name="deadzone_size" value="30" />
			<param name="left_bang_effort" value="-100" />
			<param name="right_bang_effort" value="100" />
		</node>

	</group>

</launch>
